clouddq/classes/bigquery_client.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import logging
+import re
+
+from pathlib import Path
+from string import Template
+
+import google.auth
+import google.auth.transport.requests
+
+from google.api_core.exceptions import Forbidden
+from google.api_core.exceptions import NotFound
+from google.auth import impersonated_credentials
+from google.auth.credentials import Credentials
+from google.cloud import bigquery
+from google.oauth2 import id_token
+from google.oauth2 import service_account
+
+
+logger = logging.getLogger(__name__)
+
+TARGET_SCOPES = [
+    "https://www.googleapis.com/auth/cloud-platform",
+    "https://www.googleapis.com/auth/userinfo.email",
+]
+
+CHECK_QUERY = Template(
+    """
+SELECT
+    *
+FROM (
+    $query_string
+) q
+"""
+)
+
+RE_EXTRACT_TABLE_NAME = ".*Not found: Table (.+?) was not found in.*"
+
+
+class BigQueryClient:
+    __client: bigquery.client.Client = None
+    __credentials: Credentials = None
+    __project_id: str = None
+    __user_email: str = None
+    target_audience = "https://bigquery.googleapis.com"
+
+    def __init__(
+        self,
+        credentials: Credentials = None,
+        project_id: str = None,
+        gcp_service_account_key_path: Path = None,
+        gcp_impersonation_credentials: str = None,
+    ) -> None:
+        # Use Credentials object directly if provided
+        if credentials:
+            source_credentials = credentials
+        # Use service account json key if provided
+        elif gcp_service_account_key_path:
+            source_credentials = service_account.Credentials.from_service_account_file(
+                filename=gcp_service_account_key_path, scopes=TARGET_SCOPES
+            )
+        # Otherwise, use Application Default Credentials
+        else:
+            source_credentials, _ = google.auth.default(
+                scopes=TARGET_SCOPES, quota_project_id=project_id
```

Comment:
if project id is not provided , it should use current project as default or no?

Reply:
I don't think the quota_project_id is automatically set to the project_id